plonky2x/src/frontend/builder/mux.rs

```diff
@@ -9,7 +9,22 @@ trait MuxBuilder {
 
 impl<F: RichField + Extendable<D>, const D: usize> MuxBuilder for CircuitBuilder<F, D> {
     fn select_index<T: CircuitVariable>(&mut self, selector: Variable, inputs: &[T]) -> T {
-        todo!()
+        let num_var_in_t = inputs[0].variables().len();
+        let input_len = inputs.len();
+        let mut res = (0..num_var_in_t).map(|_| self.init::<Variable>()).collect::<Vec<_>>();
+        let api = &mut self.api;
+
+        for i in 0..input_len {
+            let target_i = api.constant(F::from_canonical_usize(i));
+            let whether_select = api.is_equal(target_i, selector.0);
+
+            let vars = inputs[i].variables();
+            for j in 0..num_var_in_t {
+                res[j] = Variable(api.select(whether_select, vars[j].0, res[j].0));
```

Comment:
Because here we are selecting between many different variables and `api.select` is only meant for selecting between 2 variables, you should use the `random_access` gate in the `plonky2` api:

https://github.com/mir-protocol/plonky2/blob/main/plonky2/src/gadgets/random_access.rs#L14

If you plug this into this PR, I think you should be all set!

Reply:
yeah, think that's how it's used here? not fully get how `random_access` can be plugged here since it applies to vec<target> but not vec<vec<target>>

Reply:
Here is the pseudocode:

```
final_vars = Vec::new()
for i in 0..num_var_it_t {
   pos_vars = Vec::new() // all the variables in the `i`-th position for each element of input
   for j in 0..inputs.len() {
      pos_vars.push(inputs[j].variables()[i])
   }
   final_vars.push(api.random_access(selector, pos_vars));
}

return T::from_variables(final_vars)
```

let me know if the above doesn't make sense